Update dependencies, depend on full rspec only in development

ojab opened the pull request.

ojab commented:
Looks like current RSpec is broken for jruby-18mode, do we care about it or jruby-18mode can be dropped?

ojab commented:
Updated the PR so it applies correctly to the current master.

jfelchner commented on fuubar.gemspec:
```diff
@@ -20,7 +20,7 @@ Gem::Specification.new do |spec|
 
   spec.add_dependency             'rspec', ["~> 3.0"]
-  spec.add_dependency             'ruby-progressbar', ["~> 1.4"]
+  spec.add_dependency             'ruby-progressbar', ["~> 1.7"]
```
@ojab why do this? 1.4 works perfectly fine with the latest `fuubar`

ojab replied:
Updated testsuite is failing with 1.4, so further changes can break fuubar w/ ruby-progressbar < 1.7 and there is no way to check.
I assume that it's better to bump version than write conditional tests. Also ruby-progressbar-1.7.0 was released 1,5+ years ago, so I doubt that this change will affect any current fuubar users.

jfelchner replied:
@ojab this change has nothing to do with which `ruby-progressbar` runs during tests.  The `Gemfile.lock` does and it's set to `ruby-progressbar` `1.8`.

jfelchner commented:
@ojab we'll support JRuby in 1.8 mode as long as RSpec is compatible with Ruby 1.8.

ojab commented:
Aha, looks like copying rspec's version matrix is the right choice then. Pushed the updated travis config.

ojab commented:
JRuby fails on
```ruby
describe Object do
  it 'whatever' do
    RSpec::Core::ExampleGroup.describe.example
  end
end
```
trying to find workaround.

ojab commented:
Added workaround, as a side effect it fixed `WARNING: Your progress bar is currently at 21 out of 21 and cannot be incremented. In v2.0.0 this will become a ProgressBar::InvalidProgressError.` warnings.

jfelchner commented:
@ojab you are amazing!  Thank you for looking at this! 😀 

Any idea what that last failure is on Travis?

ojab commented:
Looks like it's https://github.com/bundler/bundler/issues/4586. Pushed version with workaround, hope this time travis will be green.

ojab commented:
Actually there is an issue with any command that runs via `bundle exec command` on bundler-1.12 & ruby-1.8, so I've added rake task for running rspec via `bundle exec rake spec` which works fine. 
Also there is an issue with selection of different rake versions for different rubies in Gemfile, so I've moved all deps to gemspec. Gemfile.lock is redundant in that config (since bundler looks only on version constraints, not locked versions, for deps in gemspec), so I've removed it.

Hope that it's acceptable and not too invasive.

jfelchner commented:
@ojab this is great!  I'd like to find a way of leaving `ruby-progressbar` at `1.4` in the `gemspec` but specify the version constraint elsewhere (like the `Gemfile` so that I don't have to bump the version just to get the specs to pass.

jfelchner commented:
@ojab GREEN BUILD!  Thank you!

david50407 commented:
When will this be merged?

jfelchner commented:
@david50407 as soon as I get a chance to cherry pick only the stuff I want to merge. Is this not being merged a problem for you?

ojab commented:
The only change after this PR -- it would be possible to use fuubar w/ pre-release versions of rspec (dependency on full rspec leads to some dependency issues).
So if you're not using pre-release rspec (which was a must for rails-5 before rspec-3.5.0), this PR should not affect you at all.

david50407 commented:
My project only uses `rspec-core` but this gem will require all `rspec` (`rspec-core`, `rspec-mocks`, `rspec-expectations`), but I want to keep my deps simple

jfelchner commented:
Alright.  I've rebased, updated and merged these commits.

@ojab thank you for this PR!

In the future though, here are a couple pointers:

* Don't change too much in one PR.  This should have been two PRs.
  * Switch out `rspec` for `rspec-core`
  * Get build passing
* Don't change the fundamental way the project is structured in a PR
  * Don't remove `Gemfile.lock` if it's included in the project
  * Don't change the command used to run the tests unless there's absolutely a need to
  * Don't change dependencies (or for that matter anything in the `gemspec`) unless that is the only change in the PR)

That said, the project is improved now because of you.  Thank you!

For some reason we're getting an intermittent failure on Travis, but I have no idea why.

jfelchner commented:
For future reference, the commits for this branch were merged in here: f6f69c23

ojab commented:
I still think that [Gemfile.lock should not be in the repo](http://yehudakatz.com/2010/12/16/clarifying-the-roles-of-the-gemspec-and-gemfile/) and `ruby-progressbar` version should be bumped to allow testing w/ all supported `ruby-progressbar` versions 😃.

Command used to run the tests was changed to make JRuby tests pass. As I've said in a comment above, it's bundler/bundler#4586 and there is no failure when tests are run via rake task (I assume that should've been a separate commit for clarity).

ojab commented:
bundler/bundler#4586 is fixed in master, so when the new bundler version will be released this issue should go away.
